Persist and republish mailbox messages

chimp1984 opened the pull request:
For better resiliance in case all seed nodes would go offline, we persist the mailbox messages and republish them at startup.
We had already a persistence for own mailbox messages and reuse that but add all mailbox messages.
We check for the TTL to not publish expired messages as well we check if we have received a remove message for that mailbox message.

To not overload too much the P2pService class I refactored the mailbox related code out to a new class. The first commits are refactoring commits. The code for the new feature starts at b91a1f7ef7ce183aa2e0dfc3899a0a954d8a7eb3

chimp1984 committed:
Refactor: Move mailbox related classed to mailbox package. Make constructor public.

chimp1984 committed:
Make some methods default

chimp1984 committed:
Move mailbox related code from P2PService to MailboxMessageService

chimp1984 committed:
Remove getMailBoxMessages from P2PService and use MailboxMessageService instead

chimp1984 committed:
Remove addDecryptedMailboxListener from P2PService and use MailboxMessageService instead

chimp1984 committed:
Remove removePrivateNotification from P2PService and use MailboxMessageService instead

chimp1984 committed:
Remove sendEncryptedMailboxMessage from P2PService and use MailboxMessageService instead

chimp1984 committed:
Cleanups

chimp1984 committed:
Add new methods, make decryptedMessageWithPubKey nullable

chimp1984 committed:
Refactor: Rename

chimp1984 committed:
Filter expired entries at startup

chimp1984 committed:
Refactor: Rename

chimp1984 committed:
Remove nullable return value from decryptProtectedMailboxStorageEntry
Remove Objects::nonNull check
Split handleMailboxItem into 2 methods for persisting and
for processing my mailbox msg

chimp1984 committed:
Add filter for isMine at getMyMailBoxMessages
Add TODOs

chimp1984 committed:
Remove old code

chimp1984 committed:
Add onRemoved method to remove mailbox entries once a remove message arrives
Add republishMailBoxMessages method

Conza88 commented:
Awesome. Yes please

chimp1984 committed:
Persist map for removed mailbox messages (AddOncePayload more generally)
We add the date when we add the hash so that we can remove expired data.

chimp1984 committed:
Refactor: move classes to persistence package

chimp1984 committed:
Refactor: rename method

chimp1984 committed:
Set addressPrefixHash to empty bytes.
We have added the capability in 1.4.0 and have
enforced 1.5.1 so no traders can use that old
version anymore so the capability check is not
needed anymore.

chimp1984 committed:
Add RemovedPayloadsStorageService
Adjust param

chimp1984 committed:
Refactor: Rename class

chimp1984 committed:
Use a hashset for listeners

chimp1984 committed:
Use MailboxItem value at mailboxItemsByUid instead of a list:
The uid is not perfect like a cryptographic hash but it can be
considered safe enough to not have collusion. The list got filled
with duplicates, so should have been a hashSet anyway.
We cannot use the hash as that is not available at the remove method.
We might refactor that in future to get rid of that problematic uid as
key but will require a but more refactoring in the client code as well,
as we do not pass around the outer envelope data but only the decrypted
data.

- Add protectedStorageEntry from persisted mailbox messages to
P2PDataStorage at startup. This ensures that we add those keys to excludedKeys
and that helps to reduce load for seed nodes at initial data response.

- Refactoring at removeFromMapAndDataStore:

- Add trace logs
- Add more comments

chimp1984 committed:
Refactor: move class to new utils package

chimp1984 committed:
Refactor: move duplicated method
capabilityRequiredAndCapabilityNotSupported to CapabilityUtils

chimp1984 committed:
Add logs

chimp1984 committed:
Set uid inside constructor

chimp1984 committed:
Add support for TTL defined by payload message so we can use
different TTL for lower prio mailbox messages like AckMessages.

As we cannot add a field without breaking signatures we
need to use the extraMap in MailboxStoragePayload

chimp1984 committed:
Add comment

chimp1984 committed:
Refactor: Rename method

chimp1984 committed:
Increase timout for BroadcastHandler to 120 sec

chimp1984 committed:
Add republishExistingProtectedMailboxStorageEntry method

Seed nodes republish the persisted mailbox messages after
startup in chunks of 50 items with a 2 minute delay.

Move check for hasSequenceNrIncreased in addProtectedStorageEntry
earlier so we return earlier. This is a very common case for return
if we receive outdated data (like republished mailbox data we have
already received).

Mark logs with ## for easier find/replace for dev testing...

chimp1984 committed:
Add temp. ignore to a failing test.

By debugging I could not see why it fails, behaviour seems correct.
I guess its a mock setup issue. Will look into it to fix it soon...

chimp1984 committed:
Improve logs, rename methods, variables

chimp1984 committed:
Refactor: return early

chimp1984 committed:
Fix wrong delay, increase timeout

sqrrm commented on p2p/src/main/java/bisq/network/p2p/mailbox/MailboxItem.java:
```diff
@@ -24,31 +24,70 @@
+        if (decryptedMessageWithPubKey != null) {
+            // We use uid from mailboxMessage in case its ours as we have the at removeMailboxMsg only the
+            // decryptedMessageWithPubKey available which contains the mailboxMessage.
+            MailboxMessage mailboxMessage = (MailboxMessage) checkNotNull(decryptedMessageWithPubKey).getNetworkEnvelope();
```
Why the need for `checkNotNull` when that's checked in the if already

chimp1984 replied:
will remove it

sqrrm commented on p2p/src/main/java/bisq/network/p2p/mailbox/MailboxMessageService.java:
```diff
@@ -406,18 +404,25 @@ private MailboxItem decryptProtectedMailboxStorageEntry(ProtectedMailboxStorageE
             e.getStackTrace();
         }
-        return null;
+        return new MailboxItem(protectedMailboxStorageEntry, null);
```
Unnecessary empty line

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/RemovedPayloadsStorageService.java:
```diff
@@ -0,0 +1,69 @@
+        long cutOffDate = System.currentTimeMillis() - MailboxStoragePayload.TTL;
+        persistenceManager.readPersisted(persisted -> {
+                    persisted.getDateByHashes().entrySet().stream()
+                            .filter(e -> e.getValue() < cutOffDate)
```
I don't think I understand the usage of this class. Here is looks like when the persisted removedPayloadsMap is read the hashes that are older than the cutoff date are stored, but that means some removed data will be lost.

chimp1984 replied:
Ups that is wrong. Should `>`. It cleans out old entries so it does not grow endlessly.
The class it to persist the remove call so we do not add again removed mailbox messages. I will add some comments as well.

sqrrm replied:
Ok, that makes sense, thought so at first, then second guessed myself and got confused.

sqrrm reviewed (commented):
Would it make sense to have seednodes request all mailbox messages on startup and then republish the discrepancy between the response and their locally persisted messages?

That would reduce the amount of republishing for the price of only seednodes asking for all messages from the other seednodes on startup.

chimp1984 committed:
Fix incorrect comparison at RemovedPayloadsService.readPersisted
Add comments
Cleanups

chimp1984 commented:
Force pushed with rebase on master and fixes from the code review.

It still needs more testing, so please do not merge yet.

chimp1984 commented:
> Would it make sense to have seednodes request all mailbox messages on startup and then republish the discrepancy between the response and their locally persisted messages?
> 
> That would reduce the amount of republishing for the price of only seednodes asking for all messages from the other seednodes on startup.

As discussed in our call I think its better to keep it more simple and to use a prog argument to enable republishing for any node (not only seed node). By default we would enable it only on one v2 and one v3 node, that is enough to increasing reliability and in any emergency case we can activate it where needed.

chimp1984 committed:
Add program argument flag republishMailboxEntries
Only republish if that flag is set, not depending if its a seedNode

chimp1984 committed:
Add null check

chimp1984 committed:
Add removeMailboxMsg with MailboxMessage as argument

chimp1984 committed:
Remove unneeded wrapping of tradeMessage into DecryptedMessageWithPubKey

chimp1984 committed:
Store PrivateNotificationMessage instead of DecryptedMessageWithPubKey

chimp1984 committed:
Refactor handling of removeMailboxMsg

chimp1984 committed:
Refactor: rename dispatchMessage to onSupportMessage (no functionality change)

chimp1984 committed:
Use removeMailboxMsg method with MailboxMessage as param type

chimp1984 committed:
Refactor handleDecryptedMessageWithPubKey to use MailboxMessage
instead of decryptedMessageWithPubKey.
Here we change a bit the behaviour as now we check also if the
TradeMessage is a MailboxMessage. As that method is only called
from MailboxMessageService domain we never get a non MailboxMessage
here anyway and it would have been a bug.

Remove removeMailboxMsg(DecryptedMessageWithPubKey decryptedMessageWithPubKey) as not used anymore

chimp1984 committed:
Refactor: rename method to make it more explicit that its only used for MailboxMsg

chimp1984 committed:
Refactor: rename getMyDecryptedMessages method to make it more explicit
that its only used for MailboxMsg

chimp1984 committed:
Refactor handling of DecryptedMessageWithPubKey

chimp1984 committed:
Inline isMyMessage methods

chimp1984 committed:
Add containsKey check

chimp1984 committed:
Clone reportedPeers to avoid ConcurrentModificationExceptions

chimp1984 committed:
Remove outdated code
Remove outdated comment
Add checkNotNull for senderNodeAddress

chimp1984 committed:
Add TradeMailboxMessage and use a TTL of 15 days
Extend TradeMailboxMessage where we had `extends TradeMessage implements MailboxMessage`

chimp1984 committed:
Let MailboxMessage extend ExpirablePayload:
For all MailboxMessage we want to be sure they have an expire date.
Add getTTL methods and TTL value.
We use 7 days for ChatMessages and 30 days for PrivateNotificationMessages.
All others use default 15 days.
We prefer to keep them explicit and not use a default method in
ExpirablePayload or MailboxMessage.
The value in PrefixedSealedAndSignedMessage will not be used as the
sender use the TTL of the payload and pass that to the MailboxStoragePayload
where we store it in the extraMap. That way we can use different TTL even the
payload message is encrypted and we would not be able to look it up.

chimp1984 committed:
Add toString

chimp1984 committed:
Refactor: rename

chimp1984 committed:
Remove comment

chimp1984 committed:
Bugfix: Add processing of SendersNodeAddressMessage to
onBundleOfEnvelopes method.

In case we received a SendersNodeAddressMessage
inside a BundleOfEnvelope we did not set the
peers address.

chimp1984 committed:
Use TradeMailboxMessage as type for getMessage in SendMailboxMessageTask

chimp1984 committed:
Refactor: rename method

chimp1984 committed:
Use MailboxMessage as param type in sendEncryptedMailboxMessage

chimp1984 committed:
Sort persisted mail box messages by age and limit total number.
Add size limit.
Add log for number of persisted mailbox msg per day

Conza88 commented:
Awesome

chimp1984 committed:
Dont log filter content if a filter got updated

chimp1984 committed:
Change prio for mailboxMessageList persitence

sqrrm commented on p2p/src/main/java/bisq/network/p2p/mailbox/MailboxMessageService.java:
```diff
@@ -271,18 +271,19 @@ public void removeMailboxMsg(MailboxMessage mailboxMessage) {
-                // call removeMailboxItemFromMap here. The onRemoved only removes foreign mailBoxMessages.
-                log.trace("## removeMailboxMsg uid={}", uid);
-                removeMailboxItemFromLocalStore(uid);
+                if (mailboxItemsByUid.containsKey(uid)) {
```
Prefer early return

sqrrm reviewed (commented):
Seems mostly good. I wonder about the size check to push out old messages though.

chimp1984 committed:
Return early

chimp1984 commented:
Pr is running on 75% of the seeds since a few days. I guess its safe now to merge into master...

sqrrm reviewed (approved):
utACK
